Modules/_uuidmodule.c

```diff
@@ -91,45 +111,1649 @@ py_windows_has_stable_node(void)
+
+    // Swap time_hi_version (bytes 6-7)
+    dst[6] = src[7];
+    dst[7] = src[6];
+
+    // Copy clock_seq and node as-is (bytes 8-15)
+    memcpy(dst + 8, src + 8, 8);
+
+    return 0;
+}
+
+static int
+validate_int(uuid_state *state, PyObject *int_value)
+{
+    if (!PyLong_Check(int_value)) {
+        PyErr_SetString(PyExc_TypeError, "value must be an integer");
+        return -1;
+    }
+
+    int cmp = PyLong_IsNegative(int_value);
+    if (cmp < 0) {
+        return -1;
+    }
+    if (cmp == 1) {
+        PyErr_SetString(PyExc_ValueError,
+            "int is out of range (need a 128-bit value)");
+        return -1;
+    }
+
+    // Check if it's greater than max (2^128 - 1)
+    cmp = PyObject_RichCompareBool(int_value, state->uint128_max, Py_GT);
+    if (cmp < 0) {
+        return -1;
+    }
+    if (cmp == 1) {
+        PyErr_SetString(PyExc_ValueError,
+            "int is out of range (need a 128-bit value)");
+        return -1;
+    }
+
+    return 0;
+}
+
+static int
+from_int(uuidobject *self, PyObject *int_value, int validate)
+{
+    // Convert a 128-bit integer to UUID bytes (big-endian)
+
+    uuid_state *state = get_uuid_state_by_cls(Py_TYPE(self));
+
+    if (validate && validate_int(state, int_value) < 0) {
+        return -1;
+    }
+
+    if (_PyLong_AsByteArray(
+            (PyLongObject *)int_value,
+            (unsigned char *)self->bytes,
+            16,
+            0,  // big-endian
+            0,  // unsigned
+            1   // with_exceptions
+        ) < 0)
+    {
+        return -1;
+    }
+
+    return 0;
+}
+
+static int
+extract_field(
+    PyObject *fields,
+    int field_num,
+    uint64_t max_value,
+    const char *error_msg,
+    uint64_t *result
+) {
+    PyObject *field = PySequence_GetItem(fields, field_num);
+    if (field == NULL) {
+        return -1;
+    }
+
+    if (!PyLong_Check(field)) {
+        PyErr_Format(PyExc_TypeError, "field %d must be an integer", field_num);
+        goto fail;
+    }
+
+    int overflow;
+    uint64_t value = PyLong_AsLongLongAndOverflow(field, &overflow);
+    if (overflow || (value == (uint64_t)-1 && PyErr_Occurred())) {
+        PyErr_Format(PyExc_ValueError, "%s", error_msg);
+        goto fail;
+    }
+
+    if (value > max_value) {
+        PyErr_Format(PyExc_ValueError, "%s", error_msg);
+        goto fail;
+    }
+
+    *result = value;
+    Py_DECREF(field);
+    return 0;
+
+fail:
+    Py_DECREF(field);
+    return -1;
+}
+
+static int
+from_fields(uuidobject *self, PyObject *fields)
+{
+    // Validate that fields is a sequence with exactly 6 elements
+    if (!PySequence_Check(fields)) {
+        PyErr_SetString(PyExc_TypeError, "fields must be a sequence");
+        return -1;
+    }
+
+    Py_ssize_t len = PySequence_Size(fields);
+    if (len != 6) {
+        PyErr_SetString(PyExc_ValueError, "fields is not a 6-tuple");
+        return -1;
+    }
+
+    #define EXTRACT_FIELD(field_num, max_value, error_msg, type, name)          \
```

Comment:
Please add do-while around your macro and dedent your define and indent the d after the # instead (`#  define`)

Reply:
I can't put do/while around the entire macro body as it defines a variable. But I can do it partially, which isn't a bad idea.